import.go

```diff
@@ -44,17 +59,107 @@ func newImporter(tree *MutableTree, version int64) (*Importer, error) {
 		return nil, errors.New("tree must be empty")
 	}
 
-	return &Importer{
-		tree:    tree,
-		version: version,
-		batch:   tree.ndb.db.NewBatch(),
-		stack:   make([]*Node, 0, 8),
-	}, nil
+	importer := &Importer{
+		tree:         tree,
+		version:      version,
+		batch:        tree.ndb.db.NewBatch(),
+		stack:        make([]*Node, 0, 8),
+		batchMtx:     sync.RWMutex{},
+		chBatch:      make(chan db.Batch, 1),
+		chBatchWg:    sync.WaitGroup{},
+		chNode:       make(chan *Node, desiredBatchSize),
+		chNodeWg:     sync.WaitGroup{},
+		chNodeData:   make(chan NodeData, desiredBatchSize),
+		chNodeDataWg: sync.WaitGroup{},
+	}
+
+	importer.chNodeWg.Add(1)
+	go serializeNode(importer)
+
+	importer.chNodeDataWg.Add(1)
+	go setBatchData(importer)
+
+	importer.chBatchWg.Add(1)
+	go batchWrite(importer)
+
+	return importer, nil
+
+}
+
+// serializeNode get the next validated and hashed node from the channel
+// and dump the node to byte buf, push the serialized node data to another channel
+func serializeNode(i *Importer) {
+	for i.batch != nil {
+		if currNode, open := <-i.chNode; open {
+			buf := bufPool.Get().(*bytes.Buffer)
+			buf.Reset()
+			if err := currNode.writeBytes(buf); err != nil {
+				panic(err)
```

Comment:
How does this change impact the application's error handling holistically? Previously we'd get an error back from `Importer.Add` if there was a problem, now we'll have a panic in a goroutine.

Reply:
Good point! I think a better solution would be to send the error into an error channel and return that through the Importer.Add function. What do you think?